src/GHC/Util.hs

```diff
@@ -325,3 +330,33 @@ wToStr :: Outputable a => W a -> String
 wToStr (W e) = showPpr baseDynFlags e
 instance Outputable a => Eq (W a) where (==) a b = wToStr a == wToStr b
 instance Outputable a => Ord (W a) where compare = compare `on` wToStr
+
+-- GHC parse trees don't contain pragmas. We work around this with
+-- (nasty) parsing of comments.
+
+-- Pragmas. Comments not associated with a span in the annotations
+-- that have the form @{-# ...#-}@.
+pragmas :: ApiAnns -> [(Located AnnotationComment, String)]
+pragmas anns =
+  [ (c, s) |
+      c@(L _ (AnnBlockComment comm)) <- fromMaybe [] $ Map.lookup noSrcSpan (snd anns)
+    , let body = trimCommentDelims comm
+    , "#" `isPrefixOf` body && "#" `isSuffixOf` body
+    , let s = trim $ take (length body - 2) (drop 1 body)
+  ]
+
+-- Flags. The first element of the pair is the (located) annotation
+-- comment that sets the flags enumerated in the second element of the
+-- pair.
+flags :: ApiAnns -> [(Located AnnotationComment, [String])]
+flags anns =
+  [(c, opts) | (c, s) <- pragmas anns, "OPTIONS_GHC" `isPrefixOf` s
+    , let opts = words $ fromJust (stripPrefix "OPTIONS_GHC" s)]
+
+-- Language extensions. The first element of the pair is the (located)
+-- annotation comment that enables the extensions enumerated by he
+-- second element of the pair.
+langExts :: ApiAnns -> [(Located AnnotationComment, [String])]
+langExts anns =
+  [(c, exts) | (c, s) <- pragmas anns, "LANGUAGE" `isPrefixOf` s
```

Comment:
Pragmas are case insensitive.

Reply:
Much better to use `Just rest <- [stripPrefix "LANGUAGE" s]` rather than adding a `fromJust` in the path. Ditto for `flags`.

Reply:
Argh. My list comprehension fu is getting there but I'm no master yet 😉

Reply:
The `Just rest <- [stripPrefix "LANGUAGE" s]` is such a nifty pattern by the way. Deep art.

Reply:
✔️